src/Containers/include/Containers/ScopeTree.h

```diff
@@ -148,6 +150,31 @@ const ScopeT* ScopeTree<ScopeT>::FindFirstChild(const ScopeT& scope) const {
   return children.begin()->second->GetScope();
 }
 
+template <typename ScopeT>
+const ScopeT* ScopeTree<ScopeT>::FindFirstVisibleScopeAfterTime(uint32_t depth,
+                                                                uint64_t min_visible_time) const {
+  auto& ordered_nodes = ordered_nodes_by_depth_.at(depth);
+
+  // First we get the first node who is fully after the provided time.
+  auto node_it = ordered_nodes.upper_bound(min_visible_time);
+
+  // Previous nodes could also be partially visible if their ending is after min_visible_time.
+  auto previous_node_it = node_it;
+  while (node_it != ordered_nodes.begin()) {
+    --previous_node_it;
+
+    if (previous_node_it->second->GetScope()->end() < min_visible_time) {
+      break;
+    }
+    node_it = previous_node_it;
+  }
```

Comment:
Nit: The while loop can be simplified as:
```cpp
  while (node_it != ordered_nodes.begin() || 
         (--previous_node_it)->second->GetScope()->end() < min_visible_time) {
   node_it = previous_node_it;
  }
```

Reply:
Sure! Modified to && and >= instead of  || and <
```
  while (node_it != ordered_nodes.begin() &&
         (--previous_node_it)->second->GetScope()->end() >= min_visible_time) {
    node_it = previous_node_it;
  }
```